src/redis_lib/mod.rs

```diff
@@ -10,44 +11,134 @@ use tracing::error;
+
+    let client = Client::open(format!("redis://{redis_hostname}:{redis_port}"))
+        .map_err(|e| {
+            error!(error = e.to_string(), "Problem creating redis client");
+        })
+        .ok()?;
+
+    let connection = match client.get_async_connection().await {
         Err(err) => {
-            error!(error=err.to_string(), "Problem creating redis client");
+            error!(error = err.to_string(), "Failed connecting to Redis");
             return None;
         }
+        Ok(con) => con,
+    };
+
+    Some((client, connection))
+}
+
+/// Increments the counter that tracks how often an email was successfully used to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification_success(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0 + 1, current.1);
+}
+
+/// Increments the counter that tracks how often an email was given to verify an account
+#[allow(unused)]
+pub async fn increment_email_used_verification(con: &mut Connection, email: &str) {
+    let email_hash = hash_email(email);
+    let current = get_email_stats(con, &email_hash).await;
+    set_email_stats(con, &email_hash, current.0, current.1 + 1);
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification_success(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.0
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.1
+}
+
+async fn set_email_stats(con: &mut Connection, email_hash: &str, successful: u64, used: u64) {
+    let command: Result<(), _> = con.set(email_hash, format!("{successful},{used}")).await;
+    if let Err(err) = command {
+        error!(
+            error = err.to_string(),
+            "Could not set stats for email {email_hash}, ({successful},{used})"
+        );
     }
-    if let Err(err) = client.get_connection() {
-        error!(error=err.to_string(), "Failed connecting to Redis");
-        return None;
+}
+
+async fn get_email_stats(con: &mut Connection, email_hash: &str) -> (u64, u64) {
+    let command: Result<Option<String>, _> = con.get(email_hash).await;
+    if let Err(err) = command {
+        error!(error = err.to_string(), "Could not read email stats");
+        return (0, 0);
+    }
+    let redis_value = command.unwrap();
+
+    if redis_value.is_none() {
+        return (0, 0);
+    }
+    let redis_value = redis_value.unwrap();
+
+    let values: Vec<&str> = redis_value.split(',').collect();
+    if values.len() != 2 {
+        error!("Redis returned invalid value for {email_hash}")
+    }
+    let used_success = values[0].parse();
```

Comment:
Use `get(index)` to read vector values to avoid panics.

Reply:
With a return above inside the if, this can't panic.

Reply:
Yes, but it is all about avoiding functions/statements that can panic at all. The idea behind rust is to write code that can't in most cases panic

Reply:
Uhh, no, that's not the point of rust at all. Safety has nothing to do with whether or not the code panics, all a panic describes is an unrecoverable error - which happen in all applications, inevitably. Safety means that the code cannot make illegal memory accesses, like use after free, or double free, or buffer overflows, which is exactly what's at fault with most security issues today. Also, this code still cannot possibly panic - using get here just introduces unnecessary complexity, and adds even more error handling, which is already most of the code in there.